### PR TITLE
Fix prototypal closure in src

### PR DESCRIPTION
**Pull request by Itee**
Take in account this bug #10547 

Understand what's happen with the example below, and playing with BOOM constant.

```javascript
const BOOM = true;

// Foo constructor
function Foo () {
}

// Bar constructor
function Bar () {
}

// Baz constructor
function Baz () {
}

// Qux constructor
function Qux () {
}

// Baz object prototype is declared BEFORE Foo prototypal closure !
Baz.prototype = {
	sayIt: function () {

		console.log( 'I\'m a Baz !' );

	}
}

// Foo create a closure with an iffe from now...
Foo.prototype = {
	whatTheySay: function () {

		var bar = new Bar()
		var baz = new Baz()
		var qux = (BOOM) ? new Qux() : undefined

		return function whatTheySay () {
			bar.sayIt();
			baz.sayIt();

			if ( qux === undefined ) qux = new Qux();
			qux.sayIt(); // <== BOOM !

		}

	}()
}

// Bar declare an inline prototype after Foo closure
Bar.prototype.sayIt = function () {
	console.log( 'I\'m a Bar !' );
}

// Naughty Qux declare object prototype AFTER Foo prototype !!!
Qux.prototype = {
	sayIt: function () {

		console.log( 'I\'m a naughty Qux !' );

	}
}

//// Look what happen...

var foo = new Foo();
foo.whatTheySay();

//if BOOM = false
//   I'm a Bar
//   I'm a Baz
//   I'm a naughty Qux !
//
//if BOOM = true
//   I'm a Bar
//   I'm a Baz
//   TypeError: qux.sayIt is not a function
```

Now, think about the fact that three will be modular, and circular reference can happen every where !

This PR will allow to lazy load every variables in closure, with a really tiny processing effect in all sources file, and totally avoid this type of error hard to track !

Cheers, and happy new year !
Tristan.

### CODE REVIEW

**Comment by mrdoob** [comment]
I think I agree.

I would change the pattern though:

```js
var vector;
```

Instead of

```js
var vector = undefined;
```

And

```js
var vector1, vector2;
```

Instead of

```js
var vector1 = undefined;
var vector2 = undefined;
```

**Comment by TristanVALCKE** [comment]
Ok i will change pattern as soon as possible !

**Comment by TristanVALCKE** [comment]
Done ! With other fix in second PR #10564

File: src/audio/AudioListener.js [hunk 1]
```diff
@@ -84,14 +84,19 @@ AudioListener.prototype = Object.assign( Object.create( Object3D.prototype ), {
 
 	updateMatrixWorld: ( function () {
 
-		var position = new Vector3();
```
Comment (mrdoob) [inline src/audio/AudioListener.js]: I'm thinking we should make an exception in this case. This function is executed many times per frame (hot path).

**Comment by mrdoob** [comment]
Hmm, I'm having second thoughts about this PR... Many of those functions are fairly critical code... Javascript engines may stop optimising those functions due the conditional. I understand this is a good practice for robust code, but I'm not sure it really applies in this case...

**Comment by TristanVALCKE** [comment]
I know that there is some function that are hot path.
But the cost is only an comparison, which (i think) is really insignificant in view of the gain in a modular architecture !

If you need to be reassure/sure, i could wrote an benchmark before and after change, that will show the difference ? And you will be able to take your decision clearer.

Else, if you know which class/methods are hot path i can filter this fix to avoid any problematic performance regression.

**Comment by looeee** [comment]
@TristanVALCKE if the change causes the Javascript compiler to stop optimising the function, the cost would be a lot higher than just the comparison. Its hard to say how much exactly as that depends a lot on the engine and the optimisations, but it could be a factor of 10 to 100 in the worst case scenario. 

I'm far from an expert in this, but I don't immediately see why this would cause optimisation to fail. You should be able to test this with node though, it uses the same engine as Chrome (V8). 
There's some example code that you might be able to adapt to the purpose [here](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers) (see the Tooling section) - I did a very basic test and it looks like the functions will still be optimised, but I don't have time to do a more in depth and be certain at the moment.

**Comment by TristanVALCKE** [comment]
@looeee Thank for the link, really interesting !

In fact the optimization problem come from the uninitialized variable ! It break the optimization to create it "out" from the closure.

Here the tests with multiple variations:

```javascript
var Benchmark = function ( title, test ) {

	var titleDom = document.createElement( 'div' );
	titleDom.style.fontWeight = 'bold';
	titleDom.textContent = title;
	document.body.appendChild( titleDom );

	var msDom = document.createElement( 'div' );
	document.body.appendChild( msDom );

	this.run = function () {

		var start = performance.now();

		test();

		msDom.textContent = performance.now() - start;

	}

};

var benchmarks = [


	new Benchmark( 'Juste the for loop', function () {

		for ( var i = 0; i < 1000000; i ++ ) {}

	} ),

	new Benchmark( 'PreDeclare single variable', function () {

		var vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

		for ( var i = 0; i < 1000000; i ++ ) {

			vector.x = vector.y + vector.z;

		}

	} ),

	new Benchmark( 'PreDeclare multiple variable', function () {

		var vector1 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
		var vector2 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
		var vector3 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
		var vector4 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
		var vector5 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

		for ( var i = 0; i < 1000000; i ++ ) {

			vector1.x = vector2.y + vector3.z;
		}

	} ),

	new Benchmark( 'LazyLoading single with Void 0', function () {

		var vector;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(vector === void 0) vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

			vector.x = vector.y + vector.z;
		}

	} ),

	new Benchmark( 'LazyLoading single with Undefined ===', function () {

		var vector;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(vector === undefined) vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

			vector.x = vector.y + vector.z;
		}

	} ),

	new Benchmark( 'LazyLoading single with Undefined ==', function () {

		var vector;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(vector == undefined) vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

			vector.x = vector.y + vector.z;
		}

	} ),

	new Benchmark( 'LazyLoading single with Null', function () {

		var vector = null;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(!vector) vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

			vector.x = vector.y + vector.z;
		}

	} ),

	new Benchmark( 'LazyLoading single with boolean init', function () {

		var _init = true;
		var vector = undefined;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(_init){
				vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
				_init = false;
			}

			vector.x = vector.y + vector.z;

		}

	} ),

	new Benchmark( 'LazyLoading multiple variables', function () {

		var vector1, vector2, vector3, vector4, vector5;

		for ( var i = 0; i < 1000000; i ++ ) {

			if(vector1 === undefined) {
				vector1 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
				vector2 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
				vector3 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
				vector4 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
				vector5 = new THREE.Vector3( Math.random(), Math.random(), Math.random() );
			}

			vector1.x = vector2.y + vector3.z;
		}

	} ),

	new Benchmark( 'PreCreate single and test about prototype', function () {

		var vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

		for ( var i = 0; i < 1000000; i ++ ) {

			if(typeof vector.__proto__.length != 'function') vector = new THREE.Vector3( Math.random(), Math.random(), Math.random() );

			vector.x = vector.y + vector.z;

		}

	} )

];

function update () {

	for ( var i = 0, l = benchmarks.length; i < l; i ++ ) {

		benchmarks[ i ].run();

	}

	setTimeout( update, 2000 );

}

update();
```

So yes... performances are decrease, but code is safer ! Damned !!!

But wait a minute !

If you read again the first test about foo, bar, baz, qux, maybe had you notice that Bar prototype is create after closure but work in the both case... Mmmmmhhh maybe this is the solution ???

Don't set prototype with literal notation will avoid this type of problem ! 

So now the question is...
Is there an performance difference between object prototype create inline and with literal notation ?

`Foo.prototype.bar = function(){}` 
AND
`Foo.prototype = { bar: function(){} }`

The only difference that i saw in chrome debugger (cf linked image) is that the method exist under prototype of constructor but not at first level (for Qux)

forbarbazqux_prototype
.

About performance:

```javascript
var Benchmark = function ( title, test ) {

	var titleDom = document.createElement( 'div' );
	titleDom.style.fontWeight = 'bold';
	titleDom.textContent = title;
	document.body.appendChild( titleDom );

	var msDom = document.createElement( 'div' );
	document.body.appendChild( msDom );

	this.run = function () {

		var start = performance.now();

		test();

		msDom.textContent = performance.now() - start;

	}

};

var benchmarks = [

	new Benchmark( 'Foo.prototype.bar', function () {

		function Foo() {}

		Foo.prototype.Bar = function() {
			return "I'm Foo"
		}
		Foo.prototype.Bar1 = function() {
			return "I'm Foo"
		}
		Foo.prototype.Bar2 = function() {
			return "I'm Foo"
		}
		Foo.prototype.Bar3 = function() {
			return "I'm Foo"
		}
		Foo.prototype.Bar4 = function() {
			return "I'm Foo"
		}

		for ( var i = 0; i < 1000000; i ++ ) {

			var foo = new Foo()
			//					console.log(foo.Bar())

		}

	} ),

	new Benchmark( 'Baz.prototype = { qux }', function () {

		function Baz () {
		}

		Baz.prototype = {

			constructor: Baz,

			Qux: function () {
				return "I'm Foo"
			},
			Qux1: function () {
				return "I'm Foo"
			},
			Qux2: function () {
				return "I'm Foo"
			},
			Qux3: function () {
				return "I'm Foo"
			},
			Qux4: function () {
				return "I'm Foo"
			},
		}

		for ( var i = 0; i < 1000000; i ++ ) {

			var baz = new Baz()
			//					console.log(baz.Qux())

		}

	} )

];

function update () {

	for ( var i = 0, l = benchmarks.length; i < l; i ++ ) {

		benchmarks[ i ].run();

	}

	setTimeout( update, 2000 );

}

update();
```

So may be we should take 5min and think about it...
To fix this type of bug (due to not really predictable circular reference), it just require to DO NOT use literal notation to create prototype of object, and it doesn't impact performance !

@mrdoob Are you agree with that ?

**Comment by TristanVALCKE** [comment]
You can found a good explanation [here](https://developers.google.com/speed/articles/optimizing-javascript#avoiding-pitfalls-with-closures) at **Avoiding pitfalls with closures** about what's happen in this code and what work around should be use.

**Pull request closed without merging**